scala.wake

```diff
@@ -337,143 +346,142 @@ global def getScalaModuleTransIvyDeps module =
-      "classesDir" → module.moduleClassesDir.inws | JString,
-      "resources" → resourceDirs | map inws | map JString | JArray,
-      "scala" → scalaJSON,
-      Nil
-    ),
-    Nil
-  )
-
-# TODO is this necessary anymore?
-target memoizeBloopTemplate name \ module =
-  def content = bloopScalaModuleJSON module | prettyJSON
-  write (bloopTemplateName module) content
-def bloopTemplate module =
-  memoizeBloopTemplate module.getScalaModuleName module
-
-def bloopTemplateName module =
-  def name = module.getScalaModuleName
-  "{bloopDir.getPathName}/{name}.json.template"
-
-def bloopConfigName module =
+# TODO
+#  - Calculate the relative path from build.sc to workspace root instead of ../../../..
+global def genMillBuild module =
   def name = module.getScalaModuleName
-  "{bloopDir.getPathName}/{name}.json"
+  def scalaVersion = module.getScalaModuleScalaVersion
+  def stringsToPrettyString strs =
+    strs
+    | map ("    \"{_}\"")
+    | catWith ",\n"
+  def pathsToPrettyString paths =
+    paths
+    | map getPathName
+    | stringsToPrettyString
+  def srcs = module.getScalaModuleAllSourceFiles | pathsToPrettyString
+  def resources = module.getScalaModuleResourceFiles | pathsToPrettyString
+  def deps =
+    def ivyDepsJars =
+      module.getScalaModuleTransIvyDeps
+      | map (expandIvyDep scalaVersion)
+      | resolveIvyDepsImpl (Some scalaVersion)
+    def depJars =
+      module.getScalaModuleTransDeps
+      | map compileScalaModule
+    depJars ++ ivyDepsJars
+    | pathsToPrettyString
+  def pluginJars =
+    module.getScalaModuleCompilerPlugins
+    | map (expandIvyDep scalaVersion)
+    | map getIvyDepJar
+  def scalacOpts =
+    def baseOpts = module.getScalaModuleScalacOptions
+    def pluginOpts =
+      pluginJars
+      | map getPathName
+      | map ("-Xplugin:../../../{_}")
+    baseOpts ++ pluginOpts
+    | stringsToPrettyString
+  "%
+    // DO NOT MODIFY
+    // This file is autogenerated by api-scala-sifive
+    import mill._, scalalib._, os._
+    object `%{name}` extends ScalaModule {
+      def scalaVersion = "%{scalaVersion.scalaVersionToString}"
+      def sourceFiles = Seq(
+    %{srcs}
+      )
+      def depJars = Seq(
+    %{deps}
+      )
+      def resourceFiles = Seq(
+    %{resources}
+      )
+      def sources = T.sources {
+        sourceFiles.map(f => PathRef(millSourcePath / RelPath(s"../../../../$f")))
+      }
+      def unmanagedClasspath = T {
+        depJars.map(f => PathRef(millSourcePath / RelPath(s"../../../../$f")))
+      }
+      def resources = T.sources {
+        resourceFiles.map(f => PathRef(millSourcePath / RelPath(s"../../../../$f")))
+      }
+      def scalacOptions = Seq[String](
+    %{scalacOpts}
+      )
+    }
+    %"
+
+# TODO should we add Scala version to path?
+#    We might be able to get mill to do this
+def millBuildDir module =
+  "{root.getPathName}/{module.getScalaModuleName}"
+
+target millBuildFile name \ module =
```

Comment:
Do you really need to use the target subkey functionality here? I may just have PTSD from all the target subkey mismatch errors we hit, but when I look at this function, nothing here depends on the unserializable part of the ScalaModule (the `FnGeneratedSources` closure), so couldn't you memoize based on just the things that `genMillBuild` actually depends on?

You can always create a convenience function around it, but you don't need the function which accepts the `ScalaModule` itself to be memoizable. E.g.

```wake
target longOperation serializableArg1 serializableArg2 ... = ...
def myFunction unserializableArg =
  longOperation unserializableArg.getArg1 unserializableArg.getArg2
```